src/main/java/org/lsc/jndi/JndiServices.java

```diff
@@ -1180,48 +1159,74 @@ public Map<String, LscDatasets> doGetAttrsList(final String base,
 						res.put(ldapResult.getNameInNamespace(), new LscDatasets(attrsValues));
 					}
 				}
-				
-				Control[] respCtls = ctx.getResponseControls();
-				if (respCtls != null) {
-					for(Control respCtl : respCtls) {
-						if (requestPagedResults && respCtl instanceof PagedResultsResponseControl) {
-							pagedResultsResponse = ((PagedResultsResponseControl) respCtl).getCookie();
-						}
-					}
-				}
-
-				if (requestPagedResults && pagedResultsResponse != null) {
-					ctx.setRequestControls(new Control[]{
-										new PagedResultsControl(pageSize, pagedResultsResponse, Control.CRITICAL)});
-				}
 
+				pagedResultsResponse = pagination();
 			} while (pagedResultsResponse != null);
-
-			// clear requestControls for future use of the JNDI context
-			if (requestPagedResults) {
-				ctx.setRequestControls(null);
-			}
+			contextRequestControls();
         } catch (CommunicationException e) {
             // Avoid handling the communication exception as a generic one
             throw e;
         } catch (ServiceUnavailableException e) {
             // Avoid handling the service unavailable exception as a generic one
             throw e;
-		} catch (NamingException e) {
+		} catch (NamingException | IOException e) {
 			// clear requestControls for future use of the JNDI context
-			ctx.setRequestControls(null);
-			LOGGER.error(e.toString());
-			LOGGER.debug(e.toString(), e);
-			
-		} catch (IOException e) {
-			// clear requestControls for future use of the JNDI context
-			ctx.setRequestControls(null);
+			contextRequestControls();
 			LOGGER.error(e.toString());
 			LOGGER.debug(e.toString(), e);
 		}
 		return res;
 	}
 
+	/**
+	 * Obtain pagination cookie to retrieve all elements in search request.
+	 * @return paging cookie
+	 * @throws IOException
+	 * @throws NamingException
+	 */
+	public byte[] pagination() throws IOException, NamingException {
+		byte[] pagedResultsResponse = null;
+		Control[] respCtls = ctx.getResponseControls();
+		if (respCtls != null) {
+			for(Control respCtl : respCtls) {
+				if (respCtl instanceof PagedResultsResponseControl) {
+					pagedResultsResponse = ((PagedResultsResponseControl) respCtl).getCookie();
+				}
+			}
+		}
+		if (pagedResultsResponse != null) {
+			ctx.setRequestControls(new Control[]{
+					new PagedResultsControl(pageSize, pagedResultsResponse, Control.CRITICAL)});
```

Comment:
I am not sure, but here maybe you should get the previous controls, remove any previous pagedcontrol, and add the new pagedcontrol with the cookie.

Reply:
As this code was duplicated in multiple location pagination() is called instead.